autorest/retriablerequest.go

```diff
@@ -0,0 +1,77 @@
+package autorest
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+)
+
+// NOTE: the GetBody() method on the http.Request object is new in 1.8.
+//       at present we support 1.7 and 1.8 so for now the branches specific
+//       to 1.8 have been commented out.
+
+// RetriableRequest provides facilities for retrying an HTTP request.
+type RetriableRequest struct {
+	req *http.Request
+	//rc    io.ReadCloser
+	br    *bytes.Reader
+	reset bool
+}
+
+// NewRetriableRequest returns a wrapper around an HTTP request that support retry logic.
+func NewRetriableRequest(req *http.Request) *RetriableRequest {
+	return &RetriableRequest{req: req}
+}
+
+// Request returns the wrapped HTTP request.
+func (rr *RetriableRequest) Request() *http.Request {
+	return rr.req
+}
+
+// Prepare signals that the request is about to be sent.
+func (rr *RetriableRequest) Prepare() (err error) {
+	// preserve the request body; this is to support retry logic as
+	// the underlying transport will always close the reqeust body
+	if rr.req.Body != nil {
+		if rr.reset {
+			/*if rr.rc != nil {
+				rr.req.Body = rr.rc
+			} else */if rr.br != nil {
+				_, err = rr.br.Seek(0, io.SeekStart)
+			}
+			rr.reset = false
```

Comment:
Oh man, this is a bit of a nit, but I don't like the name of this variable. To me `reset` means, "still in the state of having been reset" as opposed to "this needs to be reset before use." You've documented it well, so it's not the end of the world, but I'd prefer something like, `needsRewind` or reversing its usage to fit `atStart`.